### PR TITLE
ARROW-15011: [R] Generate documentation for dplyr function bindings

### PR DESCRIPTION
**Pull request by nealrichardson**
Approach:

* `register_binding` takes an additional optional argument, `notes`, where you can list any limitations or differences in behavior between the Arrow version and the R function
* These notes are put in the `.cache` environment when the nse_funcs are built.
* New script `data-raw/docgen.R` that reads `arrow:::.cache$docs` and writes out `dplyr-funcs-docs.R` containing roxygen.
* Similarly, we pull the dplyr functions we s3_register and add them to the generated docs. Unfortunately, the notes about feature limitations aren't easily kept alongside the functions themselves because of how they're registered on load. The approach here creates a list in `arrow-package.R`, where the `.onLoad()` happens, and notes go there. 
* Docs and crossreferences are generated by roxygen2 as usual. 

I deferred filling in all of the function notes. See followup JIRAs on ARROW-17665.

### CODE REVIEW

**Comment by nealrichardson** [comment]
Current status: check fails with

```
  Missing link or links in documentation object 'arrow-dplyr-functions.Rd':
    ‘add_filename’ ‘cast’ ‘[stringr:str_like]{str_like}’
```

so need to add docs for add_filename and cast and add special casing for str_like since it is not yet released still.

Also will need to rebase and add special handling for `dplyr::across()`, since it is not handled as a nse_func.

**Comment by nealrichardson** [comment]
@github-actions crossbow submit -p pr_number=14014 preview-docs

**Comment by nealrichardson** [comment]
This is ready for review. Feel free to wordsmith the doc page and bike-shed the Rd name (maybe `acero`? should be something convenient and natural to type `?name` to view)

File: r/R/dplyr-funcs-doc.R [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+#' * [trunc][base::trunc()]
+#'
+#' ## bit64
+#' 
+#' * [as.integer64][bit64::as.integer64()]
+#' * [is.integer64][bit64::is.integer64()]
+#'
+#' ## dplyr
+#' 
+#' * [across][dplyr::across()]: only supported inside `mutate()`; purrr-style lambda functions not yet supported
```
Comment (thisisnic) [inline r/R/dplyr-funcs-doc.R:188]:
```suggestion
#' * [across][dplyr::across()]: only supported inside `mutate()`, `summarize()` and `arrange()`; purrr-style lambda functions not yet supported
```

File: r/data-raw/docgen.R [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+
+docs <- arrow:::.cache$docs
+
+# Add some functions
+
+# across() is handled by manipulating the quosures, not by nse_funcs
+docs[["dplyr::across"]] <- c(
+  "only supported inside `mutate()`;", # TODO(ARROW-17362, ARROW-17387)
+  "purrr-style lambda functions not yet supported" # TODO(ARROW-17366)
+)
```
Comment (thisisnic) [inline r/data-raw/docgen.R:126]:
```suggestion
docs[["dplyr::across"]] <- c(
  "only supported inside `mutate()`, `summarize()` and `arrange()`;", # TODO(ARROW-17387, ARROW-17389, ARROW-17390)
  "purrr-style lambda functions and use of `where()` selection helper not yet supported not yet supported" #TODO(ARROW-17366)
)
```

**Pull request merged**